src/GraphQL/Types/Composite/NonNullGraphType.cs

```diff
@@ -27,18 +27,18 @@ public NonNullGraphType(IGraphType type)
             if (type is NonNullGraphType)
```

Comment:
See `ResolvedType` setter. It's `public`. `Type` has `private` setter. Let's make `ResolvedType` setter as `internal`

Reply:
Also I would like to move 
```
       protected NonNullGraphType(Type type)
        {
            if (type == typeof(NonNullGraphType))
            {
                throw new ArgumentException("Cannot nest NonNull inside NonNull.", nameof(type));
            }

            Type = type;
        }
```
into `NonNullGraphType<T>` and make `public virtual Type Type => null`. We can set `Type` only from `NonNullGraphType<T>`. Then in `NonNullGraphType<T>` :  `public override Type Type => _type`

Reply:
After these changes the API will become less controversial.

Reply:
That has no bearing on this issue/PR.

Besides, doesn't `ResolvedType` need to be settable for any implementation of `SchemaTypes`?  I'm not sure what the point of forcing people to use `BuildNamedType` is.

Reply:
Ok. I will do another PR.

Reply:
Why forcing?

Reply:
They can use one or another constructor.

Reply:
It is impossible to write a custom `SchemaTypes` implementation if the `ResolvedType` setter is `internal`.  None of the other `ResolvedType` properties on classes are read-only for the same reason.  It should stay read/write at this time.  We should, however, design a fully read-only implementation of `ISchema` that is used during execution, so there are no inadvertent changes to the schema at runtime.

Reply:
I'm talking about `ResolvedType` on `ListGraphType` and `NonNullGraphType` and not about `FieldType.ResolvedType`.